src/app/storage/AttributeStorage.h

```diff
@@ -0,0 +1,161 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#pragma once
+
+#include <app/ConcreteAttributePath.h>
+#include <app/storage/PascalString.h>
+#include <lib/core/CHIPError.h>
+#include <lib/support/Span.h>
+
+namespace chip {
+namespace app {
+namespace Storage {
+
+/// Defines an interface for persisting attribute values.
+///
+/// Provides read/write of values (char/byte arrays or numeric values) based
+/// on a endpoint/cluster/attribute path.
+///
+/// To take into account storage compatibility, the class uses the concepts of:
+///
+///  - `Value` represents something that is writable to persistent storage
+///  - `Buffer` represents something that can be read from persistent storage
+///
+class AttributeStorage
+{
+private:
+    static ByteSpan asByteSpan(CharSpan span) { return { reinterpret_cast<const uint8_t *>(span.data()), span.size() }; }
+    static ByteSpan asByteSpan(ByteSpan span) { return span; }
+
+public:
+    /// Represents an attribute value that can be written to storage.
+    ///
+    /// In particular it has information about the data content since
+    /// historical ember-specific writes for strings encode data size as
+    /// pascal strings.
+    class Value
+    {
+    public:
+        /// Represents the value type of an element
+        ///
+        /// The reason for the definitions here are to allow backwards compatibility with
+        /// ember attribute storage and storage implementations can be free to ignore the
+        /// data types if they want to store things "as-is"
+        ///
+        /// Backward compatibility considerations:
+        ///   - ember storage uses native endianess, some other storages use little-endian always
+        ///   - ember uses pascal strings for storing data and the length of the size prefix
+        ///     is dictated by the maximum string size
+        enum class Type
+        {
+            kPrimitive,           // the data represents an integer/bool/float value. The span will contain the value
+            kRaw,                 // Raw bytes, not interpreted
+            kStringOneByteLength, // string where length is never larger than 255 characters. Data is a pascal string.
+            kStringTwoByteLength, // string where length is never larger than 0xFFFF characters. Data is a pascal string.
+        };
+
+        Value(const Value & other) = default;
+
+        template <typename T>
+        Value(ShortPascalString<T> & data) : Value({ asByteSpan(data.PascalContent()), Type::kStringOneByteLength })
+        {}
+
+        template <typename T>
+        Value(LongPascalString<T> & data) : Value({ asByteSpan(data.PascalContent()), Type::kStringTwoByteLength })
+        {}
+
+        template <typename T>
+        static Value Primitive(T & value)
```

Comment:
So this does not support "odd sized" (3, 5, 6, 7 bytes) primitives, right?  But the EmberAttributeStorageImpl seems like it's trying to?  This part is confusing me.